specs/unified_kernel_image.md

```diff
@@ -159,3 +159,41 @@ that tools staging or consuming UKIs have a common place to store and look for t
 
 The installed UKIs should have a filename `<version format specification>.efi`, i.e. the filename is left to
 implementers but must be valid for comparisons according to the [Version Format Specification](version_format_specification.md).
+
+### Auxiliary resources
+
+Auxiliary UKI resources (such as PE addons for command line extensions and similar,
+as well as systemd-sysext and systemd-confext DDIs) built centrally by distributions and
+installed via package manager should be installed into locations depending on
+whether they should be applied to all UKIs installed in the ESP, or only to a
+single specific UKI.
+
+UKI auxiliary resources that shall apply to *all* installed kernels may be
+installed into `/usr/lib/modules/uki.extra.d/`. UKI auxiliary resources that
+shall apply to *one* specific installed kernel shall be instead installed into
+`/usr/lib/modules/$UNAME/$UKI.efi.extra.d/`, where `$UNAME` is the output of
+`uname -r` of the kernel included in the UKI and `$UKI` is the name of the
+corresponding centrally built UKI with the `.efi` extension stripped.
+
+The installed UKI auxiliary resources must have a specific file extension, which
+depends on the resource type:
+* for PE addons is `*.addon.efi`
+* for sysext DDIs is `*.sysext.raw`
+* for confext DDIs is `*.confext.raw`
+
+
+### Example
+Given the following parameters:
+* `ARCH=x86_64`
+* `DISTRO=fedora`
+* `UKI="bar.efi"`
+* `UKI_KERNEL_VERSION="6.9.1-1-foo"`
+* `RESOURCE_NAME="test"`
+* The resource type is `PE addon`, meaning the file extension is `.addon.efi`
+
+This is where an UKI/auxiliary resources vendor installs the files:
+* Auxiliary resource filename is `test-bar.fedora.x86_64.addon.efi`
```

Comment:
This and the `ARCH=` and `DISTRO=` above confuse me a bit. This example is the first were this comes into play and my understanding is, that e.g. `machine-id.addon.efi` would be valid. Having the example like this would probably be easily misread, that the template should always have a form `NAME.DISTRO.ARCH.TYPE` or that `NAME` in there must refer to the name of the UKI again.

How about something like this?
> Given a UKI `bar-123.efi` that includes a kernel with the release information `6.9.1-1-foo`. Consider
> - a PE addons `machine-id` that should apply to all installed UKIs,
> - a PE addons `proprietary-driver-2000` that is specific to one installed UKI, and
> - a sysext `mysysext-1.23.47^3` that again should apply to all installed UKIs.
>
> The resulting paths would be
> - `/usr/lib/modules/uki.extra.d/machine-id.addon.efi`,
> - `usr/lib/modules/6.9.1-1-foo/bar-123.efi.extra.d/proprietary-driver-2000.addon.efi`, and
> - `/usr/lib/modules/uki.extra.d/mysysext-1.23.47^3.sysext.raw`.

to cover most bases?

Reply:
ARCH and DISTRO are actually used in a concrete implementation of this, see https://gitlab.com/cki-project/kernel-ark/-/merge_requests/2917. When @poettering told me to use "specific versioned addon file name" I thought he meant that. But I agree that `machine-id.addon.efi` is valid, the versioning proposed there is just what we are using internally when deploying an RPM, it doesn't have to apply everywhere.

Reply:
I think I will just use your proposal